enonic/src/main/resources/site/parts/comments-list/comments-list.ts

```diff
@@ -0,0 +1,230 @@
+      chain(tupled((
+        userMapping: UserMapping,
+        comments: ReadonlyArray<Content<Comment>>,
+        user: Option<User>,
+        part: CommentsListPart,
+        parentPath: string,
+      ) => {
+        return sequenceT(ioEither)(
+          renderer(createThymeleafParams(userMapping, comments, isSome(user), req.path, part.config,  parentPath)),
+          right(part.config.includeCss === 'true')
+        )
+      })),
+      fold(
+        errorResponse('commentList', true),
+        tupled((res: string, includeCss: boolean) => ok(res, includePageContributions(includeCss)))
+      )
+    )();
+}
+
+function cssLink() {
+  return `<link
+           rel="stylesheet"
+           href="${assetUrl({path: 'design-system/bundle.css'})}" />`
+}
+
+function includePageContributions (includeCss: boolean) {
+  return includeCss ? {
+    pageContributions: {
+      headEnd: cssLink()
+    }
+  } : undefined
+}
+
+export function getUserMapping(res: ReadonlyArray<Content<Comment>>): IOEither<EnonicError, UserMapping> {
+
+  const set = res.map((c) => `"${c.creator}"`).join(',')
+
+  return pipe(
+    findUsers({
+      start: 0,
+      count: 100,
+      query: set.length > 0 ? `_id IN (${set})` : ''
+    }),
+    map((u) => u.hits.reduce((acc: UserMapping, curr: User) => {
+      acc[curr.key] = curr.displayName;
+      return acc
+    }, {}))
+  )
+}
+
+interface ThymeleafParams {
+  readonly comments: ReadonlyArray<ContentTree>
+  readonly serviceUrl: string
+  readonly parentPath?: string
+  readonly pathSelf: string
+  readonly loginUrl?: string;
+  readonly logOut?: string
+  readonly isAuthenticated: boolean;
+}
+
+function createThymeleafParams(
+  userMapping: UserMapping,
+  comments: ReadonlyArray<Content<Comment>>,
+  isAuthenticated: boolean,
+  path: string,
+  partConfig: CommentsListPartConfig,
+  parentPath: string,
+): ThymeleafParams {
+  return {
+    comments: createCommentsTree(userMapping, partConfig.headingTag)(comments),
+    serviceUrl: serviceUrl('comment'),
+    parentPath: parentPath,
+    pathSelf: path,
+    loginUrl: getLoginUrl(partConfig.loginPageId),
+    logOut: logoutUrl({}),
+    isAuthenticated
+  }
+}
+function getLoginUrl(partLoginPageId?: string) {
+
+  return app.config.loginPageId ?
+    pageUrl(app.config.loginPageId) :
+    partLoginPageId ?
+    pageUrl(partLoginPageId) :
+    loginUrl({});
+}
+
+function notNullOrUndefined(str: string | undefined): boolean {
+  return (str !== undefined) && (str !== null);
+}
+
+export const byParentFirst = pipe(
+  ordBoolean,
+  contramap((c: Content<Comment>) => notNullOrUndefined(c.data.parentId))
+)
+
+export function childrenToParent(acc: ReadonlyArray<ContentTree>, curr: ContentTree): ReadonlyArray<ContentTree> {
+  if (curr.parentId) {
+    return acc.map((comment: ContentTree) => {
+      if(comment._id === curr.parentId) {
+        const children = comment.children ?? []
+        return {...comment, children: [...children, curr]} as ContentTree
```

Comment:
It might be possible (and better), to add type variables to `map` to get `ContentTree`, instead of using `as` to coerce the type.